monitor.go

```diff
@@ -0,0 +1,254 @@
+	// Create statsumConfigurer
+	statsumConfigurer := func(project string) (statsum.Config, error) {
+		res, err := auth.StatsumToken(project)
+		if err != nil {
+			return statsum.Config{}, err
+		}
+		return statsum.Config{
+			Project: res.Project,
+			BaseURL: res.BaseURL,
+			Token:   res.Token,
+			Expires: time.Time(res.Expires),
+		}, nil
+	}
+
+	// Create logger and parse logLevel
+	logger := logrus.New()
+	switch strings.ToLower(logLevel) {
+	case logrus.DebugLevel.String():
+		logger.Level = logrus.DebugLevel
+	case logrus.InfoLevel.String():
+		logger.Level = logrus.InfoLevel
+	case logrus.WarnLevel.String():
+		logger.Level = logrus.WarnLevel
+	case logrus.ErrorLevel.String():
+		logger.Level = logrus.ErrorLevel
+	case logrus.FatalLevel.String():
+		logger.Level = logrus.FatalLevel
+	case logrus.PanicLevel.String():
+		logger.Level = logrus.PanicLevel
+	default:
+		panic(fmt.Sprintf("Unsupported log-level: %s", logLevel))
+	}
+
+	// Convert tags to logrus.Fields
+	fields := make(logrus.Fields, len(tags))
+	for k, v := range tags {
+		fields[k] = v
+	}
+
+	// Declare monitor so we can reference it in OnError
+	m = &Monitor{
+		Statsum: statsum.New(project, statsumConfigurer, statsum.Options{
+			OnError: func(err error) { m.ReportWarning(err) },
+		}),
+		Entry: logrus.NewEntry(logger).WithFields(fields),
+		sentry: &sentry{
+			client:  nil,
+			project: project,
+			auth:    auth,
+		},
+	}
+
+	return
+}
+
+type sentry struct {
+	client     *raven.Client
+	m          sync.Mutex
+	project    string
+	expiration time.Time
+	auth       *tcauth.Auth
+}
+
+func (s *sentry) Client() (*raven.Client, error) {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	// Refresh sentry DSN if necessary
+	if s.expiration.Before(time.Now()) {
+		// Fetch DSN
+		res, err := s.auth.SentryDSN(s.project)
+		if err != nil {
+			return nil, err
+		}
+		// Create or update DSN for the client
+		if s.client == nil {
+			s.client, err = raven.New(res.Dsn.Secret)
+		} else {
+			err = s.client.SetDSN(res.Dsn.Secret)
+		}
+		if err != nil {
+			return nil, err
+		}
+		// Set expiration, so we remember to refresh
+		s.expiration = time.Time(res.Expires)
+	}
+
+	return s.client, nil
+}
+
+// A Monitor is responsible for collecting logs, stats and error messages.
+//
+// A monitor is a context aware object for monitoring. That is to say that a
+// Monitor is used to record metrics, write logs and report errors. When doing
+// so the Monitor object adds meta-data to the metrics, logs and errors. The
+// meta-data added is context dependent tags and prefix. These help identify
+// where a log message, metric or error originates from.
+//
+// By encapsulating the context meta-data inside the Monitor object, an
+// implementor gets a Monitor rarely needs to add tags or prefix. For example
+// a monitor will always be prefixed by plugin name before being passed to a
+// plugin, hence, it is easy trace any log message, metric or error report to
+// the plugin that it was created in.
```

Comment:
Can you reword "plugin" as "feature"? ("plugin" is the name we use in taskcluster-worker, but in generic-worker we called the equivalent thing a "feature")